### DISCUSSION TITLE
"T is string" pattern?

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#5194] [Language Ideas]
```csharp
object Method<T>() =>
    T switch
    {
        string => ...,
        int => ...,
        bool when ... => ...,
        ...
    };
```
instead of?
```csharp
object Method<T>()
{
    var type = typeof(T);
    if (type == typeof(string))
        return ...;
    if (type == typeof(int))
        return ...;
    if (type == typeof(bool) && ...)
        return ...;
    ...
}
```
```csharp
//Old 
if (typeof(T1) == typeof(string))
if (typeof(T1) == typeof(T2))
if (typeof(T1) == typeof(double) || typeof(T1) == typeof(float))
//New
if (T1 is string)
if (T1 is T2)
if (T1 is double or float)
```
```csharp
if (T is int i) //Not allow
if (T is string and bool) //Not allow
```

**Comment by alrz** [comment]
Duplicate of https://github.com/dotnet/csharplang/discussions/356

**Reply by ghost** [reply]
Thanks

**Reply by acaly** [reply]
I think there are subtle differences between pattern on `T` and pattern on `typeof(T)`. The other one is just to change equality comparison of `typeof(T)` into pattern matching, while this one is asking to manipulate the generic type parameter. See [my reply](https://github.com/dotnet/csharplang/discussions/5194#discussioncomment-1343768) below.

**Reply by ghost** [reply]
I have a method
```csharp
void Method<T>()
```
My proposal is to reduce ugly code, like this
> ```csharp
>     var type = typeof(T);
>     if (type == typeof(string))
>         return ...;
>     if (type == typeof(int))
>         return ...;
>     if (type == typeof(bool) && ...)
>         return ...;
>     ...
> ```
#356

> ```csharp
> 
> switch (typeof(T)) 
> {
>   case typeof(Foo): ... break;
>   case typeof(Bar): ... break;
> }
> ```
😕Can `typeof(Foo)` be removed?
```csharp
switch (T)
{
    case Foo: ... break;
    case Bar: ... break;
}
```
😀

**Comment by acaly** [comment]
I think this would be more useful if it can interact with generic constraints, including CRTP. For example,

```c#
bool AreEqual<T>(T a, T b)
{
    if (T is IEquatable<T>)
    {
        //It's equivalent to where T : IEquatable<T>, so we should be able to:
        return a.Equals(b);
    }
    else
    {
        //Do something else.
    }
}
```

I agree this is difficult, as it seems to require IL and runtime support. Related to another discussion dotnet/runtime#58507.

Note that there are BCL codes that would benefit from this, e.g. `EqualityComparer<T>` and `ArraySortHelper<T>`, and probably more in BCL and other libraries after the generic maths is introduced.

**EDIT**

As mentioned below, if this were to be supported, the pattern matching might better be written as
```c#
if (T is IEquatable<T2> T2)
{
    //Now T2 satisfies where T2 : IEquatable<T2>
}
```

**Reply by FaustVX** [reply]
The C# team don't like to change the type of values 'on-the-fly'.
But you could use an existing feature :
```csharp
bool AreEqual<T>(T a, T b)
{
    if ((a, b) is (IEquatable<T>, IEquatable<T>) eq)
    {
        //It's equivalent to where T : IEquatable<T>, so we should be able to:
        return eq.a.Equals(eq.b);
    }
    else
    {
        //Do something else.
    }
}
```
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgMwAERNAwjQN4U3t2GEtsd/7EAnAAoAgsgCmAUQCOqCABthAVjjKAlOoDcvPuwEjx0uYuEwJAd0bD1cGuasMbmneT4BfXTS8CAbDWAAe0CFGiNZeQUAHgAVAD5hGJoIOyTgdS9WNz12KAAzGmFhFID1GigAZ0KASQiIBAhgBQlYuLtakwamlviyiRkM7JysnJyAejHqhAByKv7UKAA3RQkYBBoEQJoLAAsJSRokkBoO+S7m1rsKrYsJGgqdwNQFDAC7xuaNwJAvUboAdho/QAdBBgXUFBVhCD0q4/p4hnoJJCJL8OCM/hwJgARLbXMASBA7WAAcyBKOBaJy+EBeUUFQkcNGCI8FHcQA)

**Reply by acaly** [reply]
Your code will box a and b when `T` is a value type. This is not very good for performance, especially when used in a performance critical component like `EqualityComparer` or something that mainly uses value types like the generic maths.

**Reply by acaly** [reply]
Actually it's not necessary to change the type of a value. It's possible to workaround if we can get a new type, by converting with `Unsafe.As`. For example,

```c#
if (T is IEquatable<T2> T2)
{
    EqualityHelper<T2>.Use(Unsafe.As<T, T2>(ref a));
}

static class EqualityHelper<T> where T : IEquatable<T>
{
    public static void Use(T val) { ... }
}
```

This kind of usage (call another generic method or use another generic type with additional constraints) is also impossible using your approach of converting to an interface. This can be quite common when working with generic maths, as many of the methods are static and must be called from a type instead of a value.

**Reply by acaly** [reply]
Another issue in your code is that you should call
```c#
eq.a.Equals(b);
```
instead of
```c#
eq.a.Equals(eq.b);
```

If you open the IL view, you can see that your code actually calls `object.Equals` instead of `IEquality<T>.Equals`.

**Reply by FaustVX** [reply]
Ok, you make some points 😄

**Reply by FaustVX** [reply]
We're both wrong,
We have to write :
```csharp
if (a is IEquatable<T> eq)
{
    //It's equivalent to where T : IEquatable<T>, so we should be able to:
    return eq.Equals(b);
}
```

[Your version](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwGYACbUgYVIG9DSHystb7H2McBOACgEEATgFMAogEcArhBQ8ArHFkBKRQG427Bp16DRk6TxhCA7lR6K4pQycpnlaguwC+60i84A2UsAD23lKR1xKRQAHgAVAD4eMNIICxjgRRc6Bw0GKAAzUh4eOK9FUigAZ2yASSCIBAhgFCFwiItyvSqausiCoTEk1LSUtLSAegHShAByEs6JKAA3aSEYBFIEb1IjAAshYVIYkFImqRba+osilaMhUiK17wkUABMvC+rape8QF37yAHZSToA6CB/CooIo8RL2T7OHoaIQgoQfRh9T6MIYAERWpzAQgQa1gAHNfnC/gi0hgfhlpEUhBD+lCnIRHEA=) (It's still `object.Equals()`)
[The new version](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKAwGYACbUgYVIG9DSHystb7H2McBOACgEEATgFMAogEcArhBQ8ArHFkBKRQG427Bp16DRk6TxhCA7lR6K4pQycpnlaguwC+60i84A2UsAD23lKR1xKRQAHgAVAD4eMNIICxjgRRc6Bw0GKAAzUh4IUigAZ1IASSCIBAhgFCFwiNIhMSTUtJS0tIB6NqKEAHJC+okoADdpIRgEUgRvUiMACyFhUhiQYtLyyurIi3ypoyFSfJnvCRQAEy89iqqJ7xAXVvIAdjqxADpSlHyeRPt75yaNIQfIR3Rgte6MDoAESm2zAQgQM1gAHM6kCXiC0hgnhlpPkhD9Wn8nIRHEA===)

**Reply by acaly** [reply]
OK, I actually didn't check :p But that's wired. I didn't know `tuple1 is (T1, T2) tuple2` will give a `tuple2` with the identical type of `tuple1`.

Maybe the boxing and `object.Equals` problems are still fine as we can rely on JIT to optimize it. But either way, it can't be used to call another method with additional constraint, which is really a missing feature in C#.

**Reply by FaustVX** [reply]
Yes, I didn't know either for the tuple pattern matching.

`object.Equals()` and `IEquatable<>.Equals()` isn't the same method,
In well written code, they should return the same output, but it's not mandatory.

**Reply by ghost** [reply]
What happened?

**Reply by Joe4evr** [reply]
@acaly Previous discussion: #662

**Reply by acaly** [reply]
> @acaly Previous discussion: #662

Thanks for the link. I tried last time but didn't find a related issue.
